From 1e58e10bdde16586bf8ff8960167d8f5864bafcd Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 11 Jun 2024 16:44:07 +0100
Subject: [PATCH] lan78xx: Read initial EEE status from DT

Add two new DT properties:
* microchip,eee-enabled  - a boolean to enable EEE
* microchip,tx-lpi-timer - time in microseconds to wait before entering
                           low power state

Signed-off-by: Phil Elwell <phil@raspberrypi.com>

lan78xx: Fix boot-time RTNL warning

lan78xx_set_eee calls phylink_ethtool_set_eee, which WARNs if rtnl_lock
has not been called. This is fine when called via the ethtool_ops
.set_eee entry point, presumably because the lock is already held, but
lan78xx_probe also calls lan78xx_set_eee via lan78xx_phy_init, which
requires an explicit rtnl_lock/unlock pair to avoid the warning.

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/net/usb/lan78xx.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

--- a/drivers/net/usb/lan78xx.c
+++ b/drivers/net/usb/lan78xx.c
@@ -2930,6 +2930,29 @@ static int lan78xx_phy_init(struct lan78
 		goto phylink_uninit;
 	}
 
+	if (of_property_read_bool(phydev->mdio.dev.of_node,
+				  "microchip,eee-enabled")) {
+		struct ethtool_keee edata;
+		memset(&edata, 0, sizeof(edata));
+
+		linkmode_set_bit(ETHTOOL_LINK_MODE_1000baseT_Full_BIT,
+			edata.advertised);
+		linkmode_set_bit(ETHTOOL_LINK_MODE_100baseT_Full_BIT,
+			edata.advertised);
+
+		edata.eee_enabled = true;
+		edata.tx_lpi_enabled = true;
+		if (of_property_read_u32(phydev->mdio.dev.of_node,
+					 "microchip,tx-lpi-timer",
+					 &edata.tx_lpi_timer))
+			edata.tx_lpi_timer = 600; /* non-aggressive */
+		rtnl_lock();
+		(void)lan78xx_set_eee(dev->net, &edata);
+		rtnl_unlock();
+
+		phy_support_eee(phydev);
+	}
+
 	ret = lan78xx_configure_leds_from_dt(dev, phydev);
 	if (ret < 0)
 		goto phylink_uninit;
